biz.aQute.launcher/src/aQute/launcher/plugin/ProjectLauncherImpl.java

```diff
@@ -417,6 +417,7 @@ public Jar executable() throws Exception {
 			String name = jar.getName();
 			try (Builder b = new Builder()) {
 				b.setProperty(Constants.JPMS_MODULE_INFO, moduleInstruction);
+				b.setProperty(Constants.MAIN_CLASS, embeddedLauncherName);
```

Comment:
In making me look at this ( :-) ), I see that this construct (building a new jar using the old jar as the input and then replacing the old jar with the new jar), loses the compression and reproducibility settings (and maybe other things like the Main-Class) from the old jar. I think you should just copy the generated module-info.class from the new jar into the old jar and continue to use the old jar.

Then you don't need to set the Main-Class in the new jar.

Reply:
Like this:

```diff
diff --git a/biz.aQute.launcher/src/aQute/launcher/plugin/ProjectLauncherImpl.java b/biz.aQute.launcher/src/aQute/launcher/plugin/ProjectLauncherImpl.java
index 3fa9cf891..3a6532670 100644
--- a/biz.aQute.launcher/src/aQute/launcher/plugin/ProjectLauncherImpl.java
+++ b/biz.aQute.launcher/src/aQute/launcher/plugin/ProjectLauncherImpl.java
@@ -314,6 +314,7 @@ public class ProjectLauncherImpl extends ProjectLauncher {
 
                builderInstrs.compression()
                        .ifPresent(jar::setCompression);
+               jar.setReproducible(getProject().getProperty(REPRODUCIBLE));
 
                String includeresource = getProject().mergeProperties(Constants.INCLUDERESOURCE);
                if (!Strings.nonNullOrEmpty(includeresource)) {
@@ -414,14 +415,12 @@ public class ProjectLauncherImpl extends ProjectLauncher {
 
                String moduleInstruction = getProject().getProperty(Constants.JPMS_MODULE_INFO);
                if (moduleInstruction != null) {
-                       String name = jar.getName();
                        try (Builder b = new Builder()) {
                                b.setProperty(Constants.JPMS_MODULE_INFO, moduleInstruction);
                                b.setJar(jar);
                                b.addClasspath(jar);
-                               jar = b.build();
-                               b.removeClose(jar);
-                               jar.setName(name);
+                               jar.putResource(MODULE_INFO_CLASS, b.build()
+                                       .getResource(MODULE_INFO_CLASS));
                        }
                }
```

Reply:
interesting! I'll try this out.

Reply:
hey @bjhargrave when I do this, I get the following exception:
```
[ERROR] Failed to execute goal biz.aQute.bnd:bnd-export-maven-plugin:6.3.0-SNAPSHOT:export (export) on project com.github.rotty3000.cdi-example-owb: Already closed com.github.rotty3000.examples-cdi -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal biz.aQute.bnd:bnd-export-maven-plugin:6.3.0-SNAPSHOT:export (export) on project com.github.rotty3000.cdi-example-owb: Already closed com.github.rotty3000.examples-cdi
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:972)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.MojoExecutionException: Already closed com.github.rotty3000.examples-cdi
    at aQute.bnd.maven.export.plugin.ExportMojo.execute (ExportMojo.java:142)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:972)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.RuntimeException: Already closed com.github.rotty3000.examples-cdi
    at aQute.bnd.osgi.Jar.check (Jar.java:1280)
    at aQute.bnd.osgi.Jar.write (Jar.java:543)
    at aQute.bnd.osgi.JarResource.write (JarResource.java:32)
    at aQute.bnd.maven.export.plugin.ExportMojo.lambda$0 (ExportMojo.java:185)
    at aQute.bnd.maven.lib.resolve.BndrunContainer.execute (BndrunContainer.java:179)
    at aQute.bnd.maven.export.plugin.ExportMojo.execute (ExportMojo.java:139)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:972)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

Reply:
solved it, just had to `b.removeClose(jar);`

Reply:
Yeah, my proposed change should not have removed the `b.removeClose(jar);`. I updated the diff in the  comment. Sorry.